src/pyeval.jl

```diff
@@ -2,13 +2,25 @@ const Py_single_input = 256  # from Python.h
 const Py_file_input = 257
 const Py_eval_input = 258
 
-const _maindict = PyDict{String,PyObject,false}(PyNULL()) # cache of __main__ module dictionary
-function maindict()
-    if ispynull(_maindict.o)
-        _maindict.o = pyincref(@pycheckn ccall((@pysym :PyModule_GetDict), PyPtr, (PyPtr,), pyimport("__main__")))
+const _namespaces = Dict{Module,PyDict{String,PyObject,true}}()
+
+pynamespace(m::Module) =
+    get!(_namespaces, m) do
+        if m === Main
+            return PyDict{String,PyObject,true}(pyincref(@pycheckn ccall((@pysym :PyModule_GetDict), PyPtr, (PyPtr,), pyimport("__main__"))))
+        else
+            ns = PyDict{String,PyObject}()
+            # In Python 2, it looks like `__builtin__` (w/o 's') must
+            # exist at module namespace.  See also:
+            # http://mail.python.org/pipermail/python-dev/2001-April/014068.html
+            # https://github.com/ipython/ipython/blob/512d47340c09d184e20811ca46aaa2f862bcbafe/IPython/core/interactiveshell.py#L1295-L1299
+            PyObject(ns).setdefault("__builtin__", builtin)
```

Comment:
Should this be only `if pyversion < v"3"` then?

Reply:
I didn't put the conditional because current IPython (which supports only Python 3) is still have this hack https://github.com/ipython/ipython/blob/512d47340c09d184e20811ca46aaa2f862bcbafe/IPython/core/interactiveshell.py#L1295-L1299 But that was probably a historical effect in IPython codebase.